prow/config/config.go

```diff
@@ -659,6 +662,57 @@ type Deck struct {
 	// accepts a key of: `org/repo`, `org` or `*` (wildcard) to define what GitHub org (or repo) a particular
 	// config applies to and a value of: `RerunAuthConfig` struct to define the users/groups authorized to rerun jobs.
 	RerunAuthConfigs RerunAuthConfigs `json:"rerun_auth_configs,omitempty"`
+	// RestrictStoragePaths restricts artifact requests to specific buckets.
+	// By default, buckets listed in the GCSConfiguration are automatically allowed.
+	// Additional locations can be allowed via `AdditionalAllowedBuckets` fields.
+	RestrictStoragePaths *bool `json:"restrict_storage_paths,omitempty"`
+	// AdditionalAllowedBuckets is a list of storage buckets to allow in artifact requests
+	// (in addition to those listed in the GCSConfiguration).
+	// Setting this field requires "RestrictStoragePaths" also be set to `true`.
+	AdditionalAllowedBuckets []string `json:"additional_allowed_buckets,omitempty"`
+}
+
+// Validate performs validation and sanitization on the Deck object.
+func (d *Deck) Validate() error {
+	if len(d.AdditionalAllowedBuckets) > 0 && !d.ShouldValidateStoragePaths() {
+		return fmt.Errorf("deck.restrict_storage_paths is not enabled despite deck.additional_allowed_buckets being configured: %v", d.AdditionalAllowedBuckets)
+	}
+
+	// TODO(@clarketm): Remove "rerun_auth_config" validation in July 2020
+	if d.RerunAuthConfig != nil {
+		logrus.Warning("rerun_auth_config will be deprecated in July 2020, and it will be replaced with rerun_auth_configs['*'].")
+
+		if d.RerunAuthConfigs != nil {
+			return errors.New("rerun_auth_config and rerun_auth_configs['*'] are mutually exclusive")
+		}
+
+		d.RerunAuthConfigs = RerunAuthConfigs{"*": *d.RerunAuthConfig}
+	}
+
+	// Note: The RerunAuthConfigs logic isn't deprecated, only the above RerunAuthConfig stuff is
+	if d.RerunAuthConfigs != nil {
+		for k, config := range d.RerunAuthConfigs {
+			if err := config.Validate(); err != nil {
+				return fmt.Errorf("rerun_auth_configs[%s]: %v", k, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ShouldValidateStoragePaths returns whether or not the Deck's storage path should be validated.
+// Validation could be either disabled by default or explicitly turned off.
+func (d *Deck) ShouldValidateStoragePaths() bool {
+	if d.RestrictStoragePaths == nil {
+		return RestrictStoragePathsByDefault
```

Comment:
What is the reason of this additional indirection with the constant? Nothing except this func uses the constant or did I miss it? Furthermore I think nothing _should_ use this constant, if anything cares about this setting it should use this func instead.

Reply:
I started with just a constant, then refactored in the function later. I think my (vaguely intuitive) "reasons" for leaving both was to make it easier to change-then-delete in the future... `RestrictStoragePathsByDefault` is defined directly beside the `Deck` definition and named extremely similar to the `deck.RestrictStoragePaths` field. Toggling a bool named "ByDefault" would also feel "safer" than re-evaluating the contents of a method like "ShouldValidateSomething". I agree that consumers should be using the function over the constant.

But I'm admittedly _not_ a go-lang expert and I'm guessing/realizing that the additional layers/abstractions aren't idiomatic Go. I'll happily remove them, per your feedback.

Reply:
Removed the extra `RestrictStoragePathsByDefault` const.